src/server/http_transport.rs

```diff
@@ -0,0 +1,340 @@
+use util::{HandyRwLock, TryInsertWith};
+use kvproto::msgpb::{Message, MessageType};
+use super::http_client::{Client, Url};
+use super::http_server::V1_MSG_PATH;
+use super::http::OnResponseResult;
+use super::transport::RaftStoreRouter;
+use pd::PdClient;
+use raftstore::store::Transport;
+use raft::SnapshotStatus;
+use kvproto::raftpb::MessageType as RaftMessageType;
+use kvproto::raft_serverpb::RaftMessage;
+use raftstore::Result as RaftStoreResult;
+
+// TODO, after pd supports HTTP, this interface will be changed.
+pub trait StoreAddrPeeker: Send {
+    fn get_address(&mut self, store_id: u64) -> Result<&str>;
+}
+
+// TODO: use Http for pd later.
+pub struct PdStoreAddrPeeker<T: PdClient> {
+    cluster_id: u64,
+    pd_client: Arc<RwLock<T>>,
+    store_addrs: HashMap<u64, String>,
+}
+
+impl<T: PdClient> PdStoreAddrPeeker<T> {
+    pub fn new(cluster_id: u64, pd_client: Arc<RwLock<T>>) -> PdStoreAddrPeeker<T> {
+        PdStoreAddrPeeker {
+            cluster_id: cluster_id,
+            pd_client: pd_client,
+            store_addrs: HashMap::new(),
+        }
+    }
+}
+
+impl<T: PdClient> StoreAddrPeeker for PdStoreAddrPeeker<T> {
+    fn get_address(&mut self, store_id: u64) -> Result<&str> {
+        // TODO: do we need re-update the cache sometimes?
+        // Store address may be changed?
+        let pd_client = self.pd_client.clone();
+        let cluster_id = self.cluster_id;
+        let s = try!(self.store_addrs.entry(store_id).or_try_insert_with(|| {
+            pd_client.rl()
+                     .get_store(cluster_id, store_id)
+                     .and_then(|s| {
+                         let addr = s.get_address().to_owned();
+                         // In some tests, we use empty address for store first,
+                         // so we should ignore here.
+                         // TODO: we may remove this check after we refactor the test.
+                         if addr.len() == 0 {
+                             return Err(box_err!("invalid empty address for store {}", store_id));
+                         }
+                         Ok(addr)
+                     })
+        }));
+
+        Ok(s)
+    }
+}
+
+enum Task {
+    Msg {
+        msg: Message,
+    },
+    OnMsgResp {
+        store_id: u64,
+    },
+    Snapshot {
+        msg: Message,
+    },
+
+    Quit,
+}
+
+// Only for sending raft messages.
+struct StorePending {
+    pub url: Url,
+    pub is_posting: bool,
+    pub msgs: VecDeque<Message>,
+}
+
+impl StorePending {
+    pub fn new(url: Url) -> StorePending {
+        StorePending {
+            url: url,
+            is_posting: false,
+            msgs: VecDeque::new(),
+        }
+    }
+}
+
+struct Worker<A: StoreAddrPeeker, R: RaftStoreRouter + 'static> {
+    client: Client,
+
+    raft_router: Arc<RwLock<R>>,
+    store_peeker: A,
+
+    pub tx: Sender<Task>,
+    rx: Receiver<Task>,
+
+    store_pendings: HashMap<u64, StorePending>,
+}
+
+impl<A, R> Worker<A, R>
```

Comment:
util worker can only send task in one thread, but I need to support in multi threads.

If you support this feature later, I will change it.